pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go

```diff
@@ -0,0 +1,452 @@
+			return nil, fmt.Errorf("ControllerRegistration %q does not exist", controllerInstallation.Spec.RegistrationRef.Name)
+		}
+
+		registrationNameToInstallationName[controllerInstallation.Spec.RegistrationRef.Name] = controllerInstallation.Name
+	}
+
+	return registrationNameToInstallationName, nil
+}
+
+// deployNeededInstallations takes the list of required names of ControllerRegistrations, a mapping of ControllerRegistration
+// names to their actual objects, and another mapping of ControllerRegistration names to existing ControllerInstallations. It
+// creates or update ControllerInstallation objects for that reference the given seed and the various desired ControllerRegistrations.
+func deployNeededInstallations(
+	ctx context.Context,
+	logger *logrus.Entry,
+	c client.Client,
+	seed *gardencorev1beta1.Seed,
+	wantedControllerRegistrations sets.String,
+	controllerRegistrationNameToObject map[string]*gardencorev1beta1.ControllerRegistration,
+	registrationNameToInstallationName map[string]string,
+) error {
+	for _, registrationName := range wantedControllerRegistrations.UnsortedList() {
+		logger.Infof("Deploying wanted ControllerInstallation for %q", registrationName)
+
+		if err := deployNeededInstallation(ctx, c, seed, registrationName, controllerRegistrationNameToObject, registrationNameToInstallationName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func deployNeededInstallation(
+	ctx context.Context,
+	c client.Client,
+	seed *gardencorev1beta1.Seed,
+	registrationName string,
+	controllerRegistrationNameToObject map[string]*gardencorev1beta1.ControllerRegistration,
+	registrationNameToInstallationName map[string]string,
+) error {
+	controllerRegistration := controllerRegistrationNameToObject[registrationName]
+
+	installationSpec := gardencorev1beta1.ControllerInstallationSpec{
+		SeedRef: corev1.ObjectReference{
+			Name:            seed.Name,
+			ResourceVersion: seed.ResourceVersion,
+		},
+		RegistrationRef: corev1.ObjectReference{
+			Name:            controllerRegistration.Name,
+			ResourceVersion: controllerRegistration.ResourceVersion,
+		},
+	}
+
+	seedSpecMap, err := convertObjToMap(seed.Spec)
+	if err != nil {
+		return err
+	}
+	registrationSpecMap, err := convertObjToMap(controllerRegistration.Spec)
+	if err != nil {
+		return err
+	}
+
+	var (
+		registrationSpecHash = utils.HashForMap(registrationSpecMap)[:16]
+		seedSpecHash         = utils.HashForMap(seedSpecMap)[:16]
+
+		controllerInstallation = &gardencorev1beta1.ControllerInstallation{}
+	)
+
+	f := func() error {
+		kutil.SetMetaDataLabel(&controllerInstallation.ObjectMeta, common.SeedSpecHash, seedSpecHash)
+		kutil.SetMetaDataLabel(&controllerInstallation.ObjectMeta, common.RegistrationSpecHash, registrationSpecHash)
+		controllerInstallation.Spec = installationSpec
+		return nil
+	}
+
+	if installationName, ok := registrationNameToInstallationName[registrationName]; ok {
```

Comment:
To me this passage is a little hard to read - having the create/update operation + executing the mutating functions depending on if the ControllerInstallation already exists or not.
To me this is more clear WDYT:
``` golang
       if installationName, exists := registrationNameToInstallationName[registrationName]; exists {
		// installation already exists
		controllerInstallation.Name = installationName
	} else {
		// setting generated name as the installation does not exist yet
		controllerInstallation.GenerateName = controllerRegistration.Name + "-"
	}

	_, err = controllerutil.CreateOrUpdate(ctx, c, controllerInstallation, f)
	return err
```

Reply:
Valid point and nice suggestion, I was doing it like this initially, though, the problem was that `CreateOrUpdate` does not supporting the creation for objects with `GenerateName`. I've added small comments to explain why it is like this, hope it is clear now.